tests/unit/fem/test_pa_kernels.cpp

```diff
@@ -0,0 +1,217 @@
+}
+
+double div_non_solenoidal_field3d(const Vector &x)
+{
+   double xi = x(0);
+   double yi = x(1);
+   double zi = x(2);
+   return -cos(yi)*sin(xi)-sin(xi)*sin(zi);
+}
+
+
+double pa_divergence_testnd(int dim, void(*f1)(const Vector&, Vector&),
+                            double(*divf1)(const Vector&))
+{
+   Mesh *mesh = nullptr;
+   if (dim == 2)
+   {
+      mesh = new Mesh(2, 2, Element::QUADRILATERAL, 0, 1.0, 1.0);
+   }
+   if (dim == 3)
+   {
+      mesh = new Mesh(2, 2, 2, Element::HEXAHEDRON, 0, 1.0, 1.0, 1.0);
+   }
+
+   int order = 4;
+
+   // Vector valued
+   H1_FECollection fec1(order, dim);
+   FiniteElementSpace fes1(mesh, &fec1, dim);
+
+   // Scalar
+   H1_FECollection fec2(order, dim);
+   FiniteElementSpace fes2(mesh, &fec2);
+
+   GridFunction field(&fes1), field2(&fes2);
+
+   MixedBilinearForm dform(&fes1, &fes2);
+   dform.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+   dform.AddDomainIntegrator(new DivergenceIntegrator);
+   dform.Assemble();
+
+   // Project u = f1
+   VectorFunctionCoefficient fcoeff1(dim, f1);
+   field.ProjectCoefficient(fcoeff1);
+
+   // Check if div(u) = divf1
+   dform.Mult(field, field2);
+   FunctionCoefficient fcoeff2(divf1);
+   LinearForm lf(&fes2);
+   lf.AddDomainIntegrator(new DomainLFIntegrator(fcoeff2));
+   lf.Assemble();
+   field2 -= lf;
+
+   delete mesh;
+
+   return field2.Norml2();
+}
+
+double testfunc(const Vector &x)
+{
+   double r = cos(x(0)) + sin(x(1));
+   if (x.Size() == 3)
+   {
+      r += cos(x(2));
+   }
+   return r;
+}
+
+void grad_testfunc(const Vector &x, Vector &u)
+{
+   u(0) = -sin(x(0));
+   u(1) = cos(x(1));
+   if (x.Size() == 3)
+   {
+      u(2) = -sin(x(2));
+   }
+}
+
+double pa_gradient_testnd(int dim, double(*f1)(const Vector&),
+                          void(*gradf1)(const Vector &, Vector &))
+{
+
+   Mesh *mesh = nullptr;
+   if (dim == 2)
+   {
+      mesh = new Mesh(2, 2, Element::QUADRILATERAL, 0, 1.0, 1.0);
+   }
+   if (dim == 3)
+   {
+      mesh = new Mesh(2, 2, 2, Element::HEXAHEDRON, 0, 1.0, 1.0, 1.0);
+   }
+
+   int order = 4;
+
+   // Scalar
+   H1_FECollection fec1(order, dim);
+   FiniteElementSpace fes1(mesh, &fec1);
+
+   // Vector valued
+   H1_FECollection fec2(order, dim);
+   FiniteElementSpace fes2(mesh, &fec2, dim);
+
+   GridFunction field(&fes1), field2(&fes2);
+
+   MixedBilinearForm gform(&fes1, &fes2);
+   gform.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+   gform.AddDomainIntegrator(new GradientIntegrator);
+   gform.Assemble();
+
+   // Project u = f1
+   FunctionCoefficient fcoeff1(f1);
+   field.ProjectCoefficient(fcoeff1);
+
+   // Check if grad(u) = gradf1
+   gform.Mult(field, field2);
+   VectorFunctionCoefficient fcoeff2(dim, gradf1);
+   LinearForm lf(&fes2);
+   lf.AddDomainIntegrator(new VectorDomainLFIntegrator(fcoeff2));
+   lf.Assemble();
+   field2 -= lf;
+
+   delete mesh;
+
+   return field2.Norml2();
+}
+
+TEST_CASE("PA Divergence", "[PartialAssembly], [MixedBilinearFormBC]")
```

Comment:
I like your tests but I'm curious about "BC" in "MixedBilinearFormBC".  What does this signify?

Reply:
This is cruft that I didn't clean up. I wanted to test the boundary condition application as well but didn't complete it in the commit.